utils/s2n_ensure.c

```diff
@@ -23,5 +23,6 @@ void* s2n_ensure_memcpy_trace(void *restrict to, const void *restrict from, size
         return NULL;
     }
 
-    return memcpy(to, from, size);
+    /* use memmove instead of memcpy since it'll handle overlapping regions and not result in UB */
+    return memmove(to, from, size);
```

Comment:
This isn't related to the rest of your change. Were you seeing an issue with overlapping memory? memmove is slower than memcpy.

Reply:
When describing the safety of the `CHECKED_MEMCPY` I noticed that `memcpy` will result in UB if the ranges overlap at all. I thought it would be best for safety to use `memmove` instead so we can have a safer default.

> memmove is slower than memcpy.

Is that actually true? I would need to see evidence of that. Also, if so, how much slower? If it's comparable I would argue the increased safety is worth it.

Reply:
Hmmm reading about it, sounds like its a super minor difference on modern compilers, possibly no difference at all. If only we had performance testing set up to prove that ;)

You're probably right about switching.